rfcs/0121-hardware-driver.md

```diff
@@ -0,0 +1,88 @@
+---
+feature: Use a more canonical name for hardware driver path and modules
+start-date: 2022-02-03
+author: @jonringer
+co-authors:
+shepherd-team: @edolstra @colemickens @Ericson2314
+shepherd-leader: @colemickens
+related-issues: https://github.com/NixOS/nixpkgs/issues/141803
+---
+
+# Summary
+[summary]: #summary
+
+Currently, NixOS mounts video drivers under the path `/run/opengl-driver/`,
+but should be mounted under a more generic `/run/current-system/drivers/` path.
+The usage of opengl explicitly may have been an accurate name given the time
+in which NixOS was first created; however, graphic drivers alone include much
+more than just userland graphics libraries so this usages is now misaligned with
+current paradigms. The misalignment with this path and the related Nixpkgs
+utilities are a misnomer in most contexts, and contributes to Nixpkgs' 
+"weirdness budget".
+
+# Motivation
+[motivation]: #motivation
+
+Video drivers ship many hardware acceleration libraries these days: vulkan,
+opengl, opencl, video encoding, cuda, and many other userland libraries are
+installed. Having all of these libraries placed under a single "opengl" header
+is odd for many contributors, especially newcomers. This awkwardness is
+compounded with other specialized hardware such as fpgas, tpus, and asics
+where their function isn't related to graphics, however, also need to make
+use of a "known good" path in which the related userland libraries will be mounted.
+
+To remedy this misalignment, a new convention around placing those libraries
+should be used, preferably `/run/current-system/drivers/`.
+
+# Detailed design
+[design]: #detailed-design
+
+Most of the implementation of the current paradigms will continue, as there
+will always be a need for some "impurity" around hardware based libraries.
+The significant changes will be around naming conventions used within
+Nixpkgs and NixOS.
+
+Deprecate and move existing `/run/opengl-driver/` logic:
+- Rename `hardware.opengl` options to `hardware.drivers` (pre 22.11)
+- Rename `pkgs.addOpenGLRunpath` shell hook to `addHardwareRunpath` (pre 22.11)
+  - Alias `addOpenGLRunpath` to `addHardwareRunpath` for compatibility (pre 22.11)
+- Update nixpkgs references of `/run/opengl-driver/` to point to `/run/current-system/drivers/` (post 22.11)
```

Comment:
If `/run/current-system` is used, we should nest it under `/run/current-system/sw` at a particular suffix like `/run/current-system/sw/nix-support/drivers`. This way Nix packages can install drivers into `$out/nix-support/drivers` and if they're added to `environment.systemPackages`, they will automatically populate `/run/current-system/sw` (needs `environment.pathsToLink = [ "nix-support/drivers" ]` though.

Reply:
wouldn't this cause issues if someone were to do `environment.systemPackages = lib.mkForce [ <some minimal collection of tooling they need ];`.

Will also require changes to "driver" builds, looks like only mesa has a `driver` output currently.